File: 268. Missing Number/missingNumber.py
'''
LeetCode problem 268. Missing Number
https://leetcode.com/problems/missing-number/
'''

from typing import List


class Solution:
    def missingNumber1(self, nums: List[int]) -> int:
        """
            Time complexity = O(n^2)
            Storage complexity = O(1)
        """
        n = len(nums) + 1
        for i in range(0, n):
            if i not in nums:
                return i

    def missingNumber2(self, nums: List[int]) -> int:
        """
            Time complexity = O(n) to obtain sum_of_elements
            Storage complexity = O(1)
        """
        n = len(nums)
        sum_of_elements = sum(nums)
        actual_sum = (n * (n + 1)) / 2
        return int(actual_sum - sum_of_elements)

File: 268. Missing Number/cs/MissingNumber/MissingNumber.cs
/*
LeetCode problem 268. Missing Number
https://leetcode.com/problems/missing-number/
*/

namespace LeetCode.MissingNumber;

public class Solution
{
    public int MissingNumber1(int[] nums)
    {
        /*
        Brute force approach
        Time complexity = O(n^2)
        Space complexity = O(1)
        */
        // LeetCode guarantees nums is a valid list with at least one element
        for (int i = 0; i < nums.Length; i++)
        {
            if (!nums.Contains(i))
                return i;
        }

        return nums.Length;
    }

    public int MissingNumber2(int[] nums)
    {
        /*
        Optimal approach
        Time complexity = O(n) to obtain sumOfElements
        Space complexity = O(1)
        */
        // LeetCode guarantees nums is a valid list with at least one element
        int n = nums.Length;
        int sumOfElements = nums.Sum();
        int actualSum = (n * (n + 1)) / 2;
        return actualSum - sumOfElements;
    }
}